File: apps/desktop/src/lib/agent-diagram.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';

const STABLE_ID = /^[0-9a-f]{64}$/u;
const ARTIFACT_REF = /^[0-9a-f]{128}$/u;
const DECIMAL = /^(?:0|[1-9][0-9]{0,18})$/u;
const KINDS = ['flowchart', 'sequence', 'class', 'state', 'entityRelationship'] as const;

export type AgentDiagramKindV1 = (typeof KINDS)[number];
export type AgentDiagramExportFormatV1 = 'svg' | 'png';
export type AgentDiagramExportThemeV1 = 'light' | 'dark' | 'transparent';

export interface AgentDiagramSummaryV1 {
  artifactRef: string;
  description: string;
  kind: AgentDiagramKindV1;
  stale: boolean;
  title: string;
  userSequence: string;
}

export interface AgentDiagramArtifactV1 {
  mermaid: string;
  summary: AgentDiagramSummaryV1;
}

export type AgentDiagramArtifactsResponseV1 = {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result:
    | { kind: 'noProject' }
    | { kind: 'notFound' }
    | { artifacts: AgentDiagramSummaryV1[]; kind: 'available' };
};

export type AgentDiagramArtifactResponseV1 = {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result:
    | { kind: 'noProject' }
    | { kind: 'notFound' }
    | { artifact: AgentDiagramArtifactV1; kind: 'available' };
};

export type AgentDiagramExportResponseV1 = {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: { kind: 'cancelled' | 'exported' | 'notFound' | 'invalidPayload' | 'failed' };
};

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryAgentDiagramArtifacts(
  sessionId: string,
  userSequence: string,
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<AgentDiagramArtifactsResponseV1> {
  stable(sessionId, 'Session');
  decimal(userSequence, 'Turn');
  return parseList(
    await invokeCommand('query_agent_diagram_artifacts', {
      request: {
        protocolVersion: CURRENT_PROTOCOL_VERSION,
        sessionId,
        userSequence,
      },
    }),
  );
}

export async function queryAgentDiagramArtifact(
  sessionId: string,
  artifactRef: string,
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<AgentDiagramArtifactResponseV1> {
  stable(sessionId, 'Session');
  artifactReference(artifactRef);
  return parseArtifact(
    await invokeCommand('query_agent_diagram_artifact', {
      request: { artifactRef, protocolVersion: CURRENT_PROTOCOL_VERSION, sessionId },
    }),
  );
}

export async function exportAgentDiagram(
  input: {
    artifactRef: string;
    format: AgentDiagramExportFormatV1;
    renderedPayload: string;
    sessionId: string;
    theme: AgentDiagramExportThemeV1;
  },
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<AgentDiagramExportResponseV1> {
  stable(input.sessionId, 'Session');
  artifactReference(input.artifactRef);
  if (
    !['svg', 'png'].includes(input.format) ||
    !['light', 'dark', 'transparent'].includes(input.theme)
  )
    throw new Error('Ungültige Exportoption.');
  if (!input.renderedPayload || input.renderedPayload.length > 16 * 1024 * 1024)
    throw new Error('Das gerenderte Diagramm ist zu groß.');
  const response = obj(
    await invokeCommand('export_agent_diagram', {
      request: { ...input, protocolVersion: CURRENT_PROTOCOL_VERSION },
    }),
  );
  protocol(response.protocolVersion);
  const result = obj(response.result);
  if (
    !['cancelled', 'exported', 'notFound', 'invalidPayload', 'failed'].includes(String(result.kind))
  )
    throw new Error('Ungültige Diagramm-Exportantwort.');
  return response as AgentDiagramExportResponseV1;
}

function parseList(payload: unknown): AgentDiagramArtifactsResponseV1 {
  const response = obj(payload);
  protocol(response.protocolVersion);
  const result = obj(response.result);
  if (result.kind === 'noProject' || result.kind === 'notFound')
    return response as AgentDiagramArtifactsResponseV1;
  if (
    result.kind !== 'available' ||
    !Array.isArray(result.artifacts) ||
    result.artifacts.length > 3
  )
    throw new Error('Ungültige Diagrammliste.');
  result.artifacts.forEach(parseAgentDiagramSummaryV1);
  return response as AgentDiagramArtifactsResponseV1;
}

function parseArtifact(payload: unknown): AgentDiagramArtifactResponseV1 {
  const response = obj(payload);
  protocol(response.protocolVersion);
  const result = obj(response.result);
  if (result.kind === 'noProject' || result.kind === 'notFound')
    return response as AgentDiagramArtifactResponseV1;
  if (result.kind !== 'available') throw new Error('Ungültige Diagrammantwort.');
  const artifact = obj(result.artifact);
  parseAgentDiagramSummaryV1(artifact.summary);
  if (
    typeof artifact.mermaid !== 'string' ||
    artifact.mermaid.length > 65_536 ||
    !/^(?:flowchart TD|sequenceDiagram|classDiagram|stateDiagram-v2|erDiagram)\n/u.test(
      artifact.mermaid,
    ) ||
    artifact.mermaid.includes('<') ||
    artifact.mermaid.includes('click ') ||
    artifact.mermaid.includes('%%{')
  )
    throw new Error('Ungültiges Diagramm.');
  return response as AgentDiagramArtifactResponseV1;
}

export function parseAgentDiagramSummaryV1(value: unknown): AgentDiagramSummaryV1 {
  const item = obj(value);
  artifactReference(item.artifactRef);
  decimal(item.userSequence, 'Turn');
  if (
    !KINDS.includes(item.kind as AgentDiagramKindV1) ||
    typeof item.title !== 'string' ||
    item.title.length === 0 ||
    item.title.length > 256 ||
    typeof item.description !== 'string' ||
    item.description.length === 0 ||
    item.description.length > 2_048 ||
    typeof item.stale !== 'boolean'
  )
    throw new Error('Ungültige Diagrammübersicht.');
  return item as unknown as AgentDiagramSummaryV1;
}

function obj(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value))
    throw new Error('Ungültige Diagrammantwort.');
  return value as Record<string, unknown>;
}

function protocol(value: unknown): void {
  if (value !== CURRENT_PROTOCOL_VERSION) throw new Error('Nicht unterstützte Protokollversion.');
}

function stable(value: unknown, label: string): asserts value is string {
  if (typeof value !== 'string' || !STABLE_ID.test(value))
    throw new Error(`${label} ist ungültig.`);
}

function artifactReference(value: unknown): asserts value is string {
  if (typeof value !== 'string' || !ARTIFACT_REF.test(value))
    throw new Error('Diagramm ist ungültig.');
}

function decimal(value: unknown, label: string): asserts value is string {
  if (typeof value !== 'string' || !DECIMAL.test(value)) throw new Error(`${label} ist ungültig.`);
}

File: crates/a3-protocol/src/agent_diagram.rs
use crate::ProtocolVersion;
use serde::{Deserialize, Serialize};

/// Closed diagram family exposed to the untrusted WebView.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentDiagramKindV1 {
    /// Directed component or control flow.
    Flowchart,
    /// Time-ordered interaction between participants.
    Sequence,
    /// Static type or module relationships.
    Class,
    /// States and transitions.
    State,
    /// Entity relationships.
    EntityRelationship,
}

/// Content-safe summary of one evidence-bound diagram.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentDiagramSummaryV1 {
    /// Opaque capability bound to the active project, session, turn, and artifact.
    pub artifact_ref: String,
    /// User-message sequence that requested the diagram.
    pub user_sequence: String,
    /// Closed diagram family.
    pub kind: AgentDiagramKindV1,
    /// Bounded user-facing title.
    pub title: String,
    /// Bounded user-facing description.
    pub description: String,
    /// Whether the diagram describes an older project index.
    pub stale: bool,
}

/// One diagram plus Core-compiled Mermaid safe for local strict rendering.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentDiagramArtifactV1 {
    /// Metadata also used by list views.
    pub summary: AgentDiagramSummaryV1,
    /// Deterministic Mermaid compiled by the Core from typed elements and edges.
    pub mermaid: String,
}

/// Requests the diagrams atomically completed with one user turn.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryAgentDiagramArtifactsRequestV1 {
    protocol_version: ProtocolVersion,
    session_id: String,
    user_sequence: String,
}

impl QueryAgentDiagramArtifactsRequestV1 {
    /// Returns the protocol version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }
    /// Returns the opaque session identifier.
    #[must_use]
    pub fn session_id(&self) -> &str {
        &self.session_id
    }
    /// Returns the decimal user-message sequence.
    #[must_use]
    pub fn user_sequence(&self) -> &str {
        &self.user_sequence
    }
}

/// Result of a bounded diagram-list request.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum AgentDiagramArtifactsResultV1 {
    /// No project is active.
    NoProject,
    /// The session or turn does not exist in the active project.
    NotFound,
    /// At most three diagrams were completed for the turn.
    Available {
        /// Bounded diagram summaries.
        artifacts: Vec<AgentDiagramSummaryV1>,
    },
}

/// Versioned diagram-list response.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentDiagramArtifactsResponseV1 {
    /// Current IPC protocol version.
    pub protocol_version: ProtocolVersion,
    /// Closed result projection.
    pub result: AgentDiagramArtifactsResultV1,
}

/// Requests one artifact through a Core-issued opaque reference.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryAgentDiagramArtifactRequestV1 {
    protocol_version: ProtocolVersion,
    session_id: String,
    artifact_ref: String,
}

impl QueryAgentDiagramArtifactRequestV1 {
    /// Returns the protocol version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }
    /// Returns the opaque session identifier.
    #[must_use]
    pub fn session_id(&self) -> &str {
        &self.session_id
    }
    /// Returns the opaque artifact capability.
    #[must_use]
    pub fn artifact_ref(&self) -> &str {
        &self.artifact_ref
    }
}

/// Result of loading one diagram artifact.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum AgentDiagramArtifactResultV1 {
    /// No project is active.
    NoProject,
    /// The capability does not resolve inside the active project and session.
    NotFound,
    /// A validated artifact is available for local rendering.
    Available {
        /// Diagram metadata and Core-compiled Mermaid.
        artifact: AgentDiagramArtifactV1,
    },
}

/// Versioned single-artifact response.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentDiagramArtifactResponseV1 {
    /// Current IPC protocol version.
    pub protocol_version: ProtocolVersion,
    /// Closed result projection.
    pub result: AgentDiagramArtifactResultV1,
}

/// Native export format.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentDiagramExportFormatV1 {
    /// Sanitized scalable vector graphics.
    Svg,
    /// Validated portable network graphics.
    Png,
}

/// Visual theme used for the rendered payload.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentDiagramExportThemeV1 {
    /// Light canvas and colors.
    Light,
    /// Dark canvas and colors.
    Dark,
    /// Transparent canvas.
    Transparent,
}

/// Requests native export without accepting a destination path from the WebView.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ExportAgentDiagramRequestV1 {
    protocol_version: ProtocolVersion,
    session_id: String,
    artifact_ref: String,
    format: AgentDiagramExportFormatV1,
    theme: AgentDiagramExportThemeV1,
    rendered_payload: String,
}

impl ExportAgentDiagramRequestV1 {
    /// Returns the protocol version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }
    /// Returns the opaque session identifier.
    #[must_use]
    pub fn session_id(&self) -> &str {
        &self.session_id
    }
    /// Returns the opaque artifact capability.
    #[must_use]
    pub fn artifact_ref(&self) -> &str {
        &self.artifact_ref
    }
    /// Returns the requested output format.
    #[must_use]
    pub const fn format(&self) -> AgentDiagramExportFormatV1 {
        self.format
    }
    /// Returns the requested visual theme.
    #[must_use]
    pub const fn theme(&self) -> AgentDiagramExportThemeV1 {
        self.theme
    }
    /// Returns the bounded rendered payload to validate at the privileged boundary.
    #[must_use]
    pub fn rendered_payload(&self) -> &str {
        &self.rendered_payload
    }
}

/// Native export result without disclosing the chosen file-system path.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum AgentDiagramExportResultV1 {
    /// The user closed the native dialog.
    Cancelled,
    /// The validated artifact was written atomically.
    Exported,
    /// The capability no longer resolves in the active project and session.
    NotFound,
    /// The rendered SVG or PNG failed strict validation.
    InvalidPayload,
    /// The native write could not be completed safely.
    Failed,
}

/// Versioned diagram-export response.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct AgentDiagramExportResponseV1 {
    /// Current IPC protocol version.
    pub protocol_version: ProtocolVersion,
    /// Path-free export result.
    pub result: AgentDiagramExportResultV1,
}
